### ISSUE TITLE
Rollout reward evaluation is serial — how to parallelize LLM-based reward computation?

### ISSUE THREAD

**Issue by Win7win** [verl-project/verl#2236]
###  Problem

I'm using VERL to train a `qwen2.5-7B` model with a custom reward function that queries an external LLM API for answer evaluation.

To speed up reward computation, I launched two local vLLM API servers and rewrote the reward function to route requests in a round-robin fashion between them.

However, I observed that only one API server is ever active at a time — they alternate between steps, but are never utilized concurrently.
GPU logs confirm that `compute_score()` appears to be invoked serially per sample, rather than in parallel.

This behavior severely limits throughput and defeats the purpose of multi-endpoint deployment.

---

### Reward function snippet

```python
import itertools
from openai import AsyncOpenAI

@lru_cache(maxsize=1)
def _client_cycle():
    c1 = AsyncOpenAI(base_url="http://localhost:8000/v1", api_key="dummy")
    c2 = AsyncOpenAI(base_url="http://localhost:8001/v1", api_key="dummy")
    return itertools.cycle([c1, c2])  # Round-robin

async def _chat(messages):
    client = next(_client_cycle())
    rsp = await client.chat.completions.create(
        model="qwen7b-think",
        messages=messages,
        max_tokens=2048,
        temperature=0.7,
        stop=["</response>"],
        stream=False,
    )
    return rsp.choices[0].message.content

async def run_once(question, experts):
    prompt = f"<question>{question}</question>\n"
    messages = [{"role": "user", "content": prompt}]
    full = await _chat(messages)
    # ... parse response ...

def compute_score(solution_str, ground_truth, format_score=0.2, full_score=1.0):
    exinfos = extract_info(solution_str)
    if exinfos is None:
        return 0.0

    question, gt_ans = split_gt(ground_truth[0])

    try:
        pred_ans = asyncio.run(run_once(question, exinfos))
    except Exception:
        return format_score

    return full_score if pred_ans == gt_ans else format_score
```

---

### Feature request or question

* Is there a way to make rollout call `compute_score()` concurrently for a batch of samples?
* Alternatively, is it possible to access the entire rollout batch within the reward hook?
* Would you recommend modifying the reward engine to support a `compute_score_batch()` interface?

Any suggestions or best practices would be greatly appreciated — this is currently a significant bottleneck in our training loop. Thanks in advance!

**Comment by 0x404** [comment]
Are you using the default `naive` reward manager? which calculation is serial. There is a [`PrimeRewardManager`](https://github.com/volcengine/verl/blob/a306434806ac7c309f0e907a765fa7aca79ff651/verl/workers/reward_manager/prime.py) which support parallel computation, hope this could help.

**Comment by yyDing1** [comment]
Some additional suggestions to speed up training using remote GenRM:
- **Server Backend:** You can use the SGLang server, which supports data parallelism. This way, you only need to configure a single remote router address.
- **Parallelism:**
    - Opt 1: `compute_score_batch` + `multiprocess request` (https://github.com/volcengine/verl/pull/2121) have been supported, you can refer to `recipe/genrm_remote` for more details
    - Opt 2 (WIP): Maybe `async_rollout` + `async_genrm_reward` can be a better choice to speed up the whole training process

**Comment by george1459** [comment]
I ran into #2236 after following advice here. I made a fix ([discussion here](https://github.com/volcengine/verl/pull/1296#issuecomment-3112479426), [fix here](https://github.com/volcengine/verl/compare/main...george1459:verl:wip/fix_custom_reward_ray)). Posting here as information for anyone who might also encounter this.

**Comment by Kirrito-k423** [comment]
中文：

这里的观察是符合当前 legacy/default 路径行为的：如果使用的是 `NaiveRewardManager`，它会逐样本调用 `compute_score()`，所以即使 `compute_score()` 内部选择不同 endpoint，外层仍然不会让多个样本同时进入该函数。PR #6673 也明确提到 `NaiveRewardManager` 是 serial per-sample scoring；该 PR 只增加了可选 timeout，不能把 naive 路径变成并行。

可选路径如下：

1. 如果只是想要一个 batch hook，可以使用当前 main 上的 `BatchRewardManager`（`reward_manager=batch` 对应的实现）。它会把 `data_sources`、`solution_strs`、`ground_truths`、`extra_infos` 一次传给你的 `compute_score`，然后你可以在函数内部用 `asyncio.gather`、连接池或多 endpoint 调度并发请求外部 LLM。
2. 如果希望沿用已有并行 reward manager，可以参考 `PrimeRewardManager`，它用 process pool + async gather 做 batch 内并行评分。
3. 对 LLM/API-based reward，更推荐看新的 Reward Loop 路径：文档中的 `RewardLoopManager` 会按 `reward.num_workers` 把 batch 分发给多个 Ray reward workers，每个 worker 内部的 `compute_score_batch` 也会异步并发处理样本。文档还建议外部模型 API/sandbox 场景使用 async reward functions。
4. 如果外部 API 有并发或速率限制，`RateLimitedRewardManager` 已经提供 `max_concurrent`、`max_rpm`、`max_tpm` 和 timeout 这类控制，更适合 LLM-as-judge 场景。

因此，不建议在 serial naive 路径里只靠 round-robin client 来实现并行；应该改用 batch/prime/reward loop/rate_limited 这类会在外层调度 batch 并发的 reward manager，或者在 batch hook 内部显式并发请求多个 endpoint。

English:

Your observation is expected on the legacy/default path: `NaiveRewardManager` calls `compute_score()` sample by sample. That means choosing different endpoints inside `compute_score()` can alternate requests, but it does not make multiple samples enter the function concurrently. PR #6673 also documents that `NaiveRewardManager` performs serial per-sample scoring; it only added an optional timeout and does not make the naive path parallel.

Available options:

1. If you want a batch hook, use the current `BatchRewardManager` (`reward_manager=batch`). It passes `data_sources`, `solution_strs`, `ground_truths`, and `extra_infos` to your reward function at once, so you can implement `asyncio.gather`, connection pooling, or multi-endpoint scheduling inside that function.
2. If you want an existing parallel reward manager, `PrimeRewardManager` already uses a process pool plus async gather for batch reward scoring.
3. For LLM/API-based reward computation, the newer Reward Loop path is usually a better fit: `RewardLoopManager` splits a batch across `reward.num_workers` Ray reward workers, and each worker's `compute_score_batch` processes samples asynchronously. The docs also recommend async reward functions for external model API or sandbox-based reward computation.
4. If the external API has concurrency or rate limits, `RateLimitedRewardManager` provides `max_concurrent`, `max_rpm`, `max_tpm`, and timeout controls for LLM-as-judge style reward calls.

So the recommended fix is not only round-robin clients inside the serial naive `compute_score()`, but switching to a batch/prime/reward-loop/rate-limited reward manager, or implementing explicit concurrent calls inside a batch reward hook.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
